### ISSUE TITLE
aliBuild should recompile only packages affected by a default

### ISSUE THREAD

**Issue by ktf** [alisw/alibuild#304]
Right now when you change a default, all the packages are affected by it and therefore:

```
--defaults release
--defaults coverage
```

result in rebuilding two hierarchies of packages. This is of course an annoyance, since the "coverage" default does not affect things like autotools, gcc or even Geant4, by policy. It would be therefore useful to make sure that a default can declare it's abilities:

```
# In default-coverage.sh
provides:
  - coverage-builds
```

```
# In default-debug.sh
provides:
  - coverage-builds
```

and packages then declare which abilities they are subscribed to:

```
# In aliroot.sh
can_use:
  - coverage-builds
  - debug
```

packages which do not declare what they `can_use` are considered to be built with `--default release`.

**Comment by Barthelemy** [comment]
looking forward to it

**Comment by Barthelemy** [comment]
Something that was in #397 which might be missing in this issue is the following. If A depends on B and we build first B (`aliBuild --defaults o2-daq build B`) and then A (`aliBuild --defaults o2-daq build A`), B will be built twice. This is really annoying.

**Comment by ktf** [comment]
If confirmed that is a bug, unless you meant `--defaults o2` in one of the cases.

**Comment by Barthelemy** [comment]
I used the same defaults in both cases.

**Comment by ktf** [comment]
This is being addressed by #415 . Closing this and continuing discussion there.
